packages/common/src/cis2/util.ts

```diff
@@ -0,0 +1,504 @@
+    const amount = serializeTokenAmount(transfer.tokenAmount);
+    const from = serializeAddress(transfer.from);
+    const to = serializeReceiver(transfer.to);
+    const data = serializeAdditionalData(transfer.data ?? '');
+
+    return Buffer.concat([id, amount, from, to, data]);
+}
+
+/**
+ * Serializes a list of {@link CIS2.Transfer} data objects according to the CIS-2 standard.
+ *
+ * @param {CIS2.Transfer[]} transfers - A list of {@link CIS2.Transfer} objects
+ *
+ * @example
+ * const transfers = [{
+    tokenId: '';
+    tokenAmount: 100n;
+    from: '3nsRkrtQVMRtD2Wvm88gEDi6UtqdUVvRN3oGZ1RqNJ3eto8owi',
+    to: '3nsRkrtQVMRtD2Wvm88gEDi6UtqdUVvRN3oGZ1RqNJ3eto8owi',
+    data: '48656c6c6f20776f726c6421';
+}];
+ * const bytes = serializeCIS2Transfers(transfers);
+ */
+export const serializeCIS2Transfers = makeSerializeList(serializeCIS2Transfer);
+
+function serializeCIS2OperatorUpdate(update: CIS2.UpdateOperator): Buffer {
+    const type = encodeWord8(update.type === 'add' ? 1 : 0);
+    const address = serializeAddress(update.address);
+    return Buffer.concat([type, address]);
+}
+
+/**
+ * Serializes a list of {@link CIS2.UpdateOperator} data objects according to the CIS-2 standard.
+ *
+ * @param {CIS2.UpdateOperator[]} updates - A list of {@link CIS2.UpdateOperator} objects
+ *
+ * @example
+ * const updates = [{
+*       type: 'add',
+*       address: '3nsRkrtQVMRtD2Wvm88gEDi6UtqdUVvRN3oGZ1RqNJ3eto8owi'
+    }];
+ * const bytes = serializeCIS2OperatorUpdates(updates);
+ */
+export const serializeCIS2OperatorUpdates = makeSerializeList(
+    serializeCIS2OperatorUpdate
+);
+
+/**
+ * Serializes {@link CIS2BalanceOfQuery} data objects according to CIS-2 standard.
+ */
+function serializeCIS2BalanceOfQuery(query: CIS2.BalanceOfQuery): Buffer {
+    const token = serializeCIS2TokenId(query.tokenId);
+    const address = serializeAddress(query.address);
+    return Buffer.concat([token, address]);
+}
+
+/**
+ * Serializes a list of {@link CIS2.BalanceOfQuery} data objects according to the CIS-2 standard.
+ *
+ * @param {CIS2.BalanceOfQuery[]} queries - A list of {@link CIS2.BalanceOfQuery} objects
+ *
+ * @example
+ * const queries = [{tokenId: '', address: '3nsRkrtQVMRtD2Wvm88gEDi6UtqdUVvRN3oGZ1RqNJ3eto8owi'}];
+ * const bytes = serializeCIS2BalanceOfQueries(queries);
+ */
+export const serializeCIS2BalanceOfQueries = makeSerializeList(
+    serializeCIS2BalanceOfQuery
+);
+
+/**
+ * Deserializes response of CIS-2 balanceOf query according to CIS-2 standard.
+ *
+ * @param {HexString} value - The hex string value to deserialize
+ *
+ * @returns {bigint[]} A list of token balances.
+ */
+export const deserializeCIS2BalanceOfResponse = makeDeserializeListResponse(
+    (buf) => {
+        const end = buf.subarray(0).findIndex((b) => b < 2 ** 7) + 1; // Find the first byte with most significant bit not set, signaling the last byte in the leb128 slice.
+        const amount = uleb128.decode(Buffer.from(buf.subarray(0, end)));
```

Comment:
Consider checking that end is not `-1` and not larger than `37` (or is it 36?)

Reply:
Why would it be necessary to check the length when deserializing the response?

Reply:
First, if there are no bytes less than `2**7` it will return the token amount of `0` and this will be hard to debug for the caller.
It is good practice to fail when something unexpected is happening.

Second, if there are more than 37 bytes, it is no longer a valid token amount.
If the rest of the application assumes this to be a valid token amount, but a bug in a contract or even ill-behaving contract can make this deserialize into invalid token amounts, breaking that assumption.

If you don't want to add checks here, we should at least document that it is up to the caller to check these things.